Propagate unreliable peers upwards from the trie accumulator to the block synchronizer

alsrdn opened the pull request:
The trie accumulator uses a snapshot of the peer list provided by the block synchronizer through the global state synchronizer in order to get the required tries.

Because we didn't provide the block synchronizer any feedback on what the peers that dind't have the global state data were, the block synchronizer may re-issue the global state sync request with the same peer set as for the previous failed attempt. This may lead to delays in getting the global state data.

This PR adds:
* the ability to propagate the unreliable peers (that didn't have the requested data) upwards to the global state synchronizer and then to the block synchronizer. This gives feedback that the original peer list provided should be changed if another global sync request is made.
* block synchronizer now demotes the peers unreliable received from the global state synchronizer
* a test to verify that the peer list is changed based on the feedback from the global state sync components.
* slight change into how `PeerQuality` is handled when demoting and promoting peers; now peers with `Reliable` or `Unreliable` quality will not be set as `Unknown` when demoting or promoting respectively.

These changes reduce the risk that the global state synchronizer gets stuck asking the same peers all over again.

Further potential improvements that can be made:
* Promote peers that were successful when downloading global state. We only demote the peers now.
* Use a reputation based peer list (similar to the one used in the block synchronizer) in the trie accumulator rather than going through peers one by one.

Fixes: https://github.com/casper-network/casper-node/issues/3563

alsrdn committed:
trie_accumulator: remove unused error variant

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
block_synchronizer/peer_list: change Unknown peer behaviour

The only time a peer has `Unknown` quality is when it was just added to
the peer list and we haven't heard anything from it.
If a peer that has `Unknown` quality is promoted we now make it
`Reliable` since it means that an item we required was found on that
node.
Also if a peer with `Unknown` quality is demoted, we make it
`Unreliable`.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
block_synchronizer: propagate unreliable peers for global state sync

The trie accumulator uses a snapshot of the peer list provided by the
block synchronizer through the global state synchronizer in order to get
the required tries.
Propagate the unreliable peers (that didn't have the requested data)
upwards to the global state synchronizer and then to the block
synchronizer. This gives feedback that the original peer list provided
should be changed if another global sync request is made.
Demote the peers received from the global state synchronizer in the
block synchronizer.
Also add a test to verify that the peer list is changed based on the
feedback from the global state sync components.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn commented:
bors try

rafal-ch commented on node/src/components/block_synchronizer.rs:
```diff
@@ -887,21 +894,43 @@ impl BlockSynchronizer {
+                        error!(%block_hash, %error, "BlockSynchronizer: failed to apply global state");
+                    }
+                }
+                // Demote all the peers where we didn't find the required global state tries
+                for peer in unreliable_peers.iter() {
+                    builder.demote_peer(Some(*peer));
+                }
```
Shouldn't we demote peers unconditionally (i.e., before the `if let` in [line 921](https://github.com/casper-network/casper-node/pull/3594/files#diff-ad7e9cbef898bd82de2888993921f091b4dbcec0c2445eb40ea244873b296aedR921))?

alsrdn replied:
Global state should be synced only on historical blocks and we'll still need to match the block hash in the response to the one from the builder because the builder provides the peers in this case.

rafal-ch commented on node/src/components/block_synchronizer/peer_list.rs:
```diff
@@ -110,13 +110,10 @@ impl PeerList {
                 }
                 Entry::Occupied(entry) => match entry.get() {
-                    PeerQuality::Dishonest | PeerQuality::Unknown => {
+                    PeerQuality::Dishonest | PeerQuality::Unreliable => {
```
Can you consider adding a UT for `demote_peer()`, `promote_peer()`, et al., just to have a clear picture of how the functions behave.

alsrdn replied:
Yes, good idea!

alsrdn replied:
Added in https://github.com/casper-network/casper-node/pull/3594/commits/1122fd2b19b2adcc12ecedf7c8d00363146ca201

rafal-ch commented on node/src/components/block_synchronizer/peer_list.rs:
```diff
@@ -148,34 +145,40 @@ impl PeerList {
-        // get most useful up to limit
-        let mut peers: Vec<NodeId> = self
-            .peer_list
+    fn get_peers_by_quality(
```
```suggestion
    fn get_random_peers_by_quality(
```

alsrdn replied:
Changed in https://github.com/casper-network/casper-node/pull/3594/commits/1122fd2b19b2adcc12ecedf7c8d00363146ca201

rafal-ch commented on node/src/components/block_synchronizer/peer_list.rs:
```diff
@@ -148,34 +145,40 @@ impl PeerList {
+        // peer list
+        let missing = up_to.saturating_sub(peers.len());
+        if missing > 0 {
+            peers.extend(self.get_peers_by_quality(rng, up_to, PeerQuality::Unreliable));
```
Should we also try to get peers of `Unknown` quality here?

alsrdn replied:
Yes, the intention was to first get `Reliable` peers, then if those are not enough extend the peers list and get `Unknown` peers and even if then we don't have enough peers, get unreliable ones. But I did a mistake here and used `up_to` in all of the calls instead of using `missing`. Good catch! I'll fix this.

alsrdn replied:
Fixed in https://github.com/casper-network/casper-node/pull/3594/commits/1122fd2b19b2adcc12ecedf7c8d00363146ca201

rafal-ch commented on node/src/components/block_synchronizer/tests.rs:
```diff
@@ -0,0 +1,212 @@
+    global_sync_request.peers
+}
+
+#[tokio::test]
+async fn global_state_sync_wont_stall_with_bad_peers() {
```
Nice test! :+1:

alsrdn committed:
block_synchronizer/peer_list: fix number of qualified peers

Fix a bug that was pointed out in code review related to the number of
peers returned when getting qualified peers.
Also add some unit tests for the peer list.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

rafal-ch reviewed (approved):
LGTM :+1:

Fraser999 commented on node/src/components/block_synchronizer/global_state_synchronizer.rs:
```diff
@@ -32,21 +32,44 @@ const COMPONENT_NAME: &str = "global_state_synchronizer";
-    TrieAccumulator(TrieAccumulatorError),
-    #[error("ContractRuntime failed to put a trie into global state: {0}")]
-    PutTrie(engine_state::Error),
+    #[error("trie accumulator encountered an error while fetching a trie; unreliable peers {0:?}")]
```
I think we should avoid mixing debug output into display output.  There's `utils::DisplayIter` which could probably help here.

Same comment for the `PutTrie` variant below.

alsrdn replied:
Fixed in https://github.com/casper-network/casper-node/pull/3594/commits/18b802deab67b7404bb7570a085b695e73a9b8f1

Fraser999 reviewed (commented):
One minor change requested, otherwise LGTM.

alsrdn committed:
global state sync: use Display format for error messages

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
Merge branch 'dev' into global-state-sync-bug

EdHastingsCasperAssociation commented on node/src/components/block_synchronizer/global_state_synchronizer.rs:
```diff
@@ -25,28 +25,52 @@ use crate::{
     NoPeersAvailable(Digest),
 }
 
+#[derive(Debug, Clone)]
+pub(crate) struct Response {
+    hash: Digest,
+    unreliable_peers: Vec<NodeId>,
+}
+
+impl Response {
+    pub(crate) fn new(hash: Digest, unreliable_peers: Vec<NodeId>) -> Self {
+        Self {
+            hash,
+            unreliable_peers,
+        }
+    }
+
+    pub(crate) fn hash(&self) -> &Digest {
+        &self.hash
+    }
+
+    pub(crate) fn unreliable_peers(self) -> Vec<NodeId> {
+        self.unreliable_peers
+    }
+}
+
```
you could do a two state enum for Success / Failure, and flatten the Error as a field on the failure variant; this would allow places that return Result<Response, Error> to be simplified to just return the enum variant (etc).

alsrdn replied:
Yes, we could to that though I'm more partial to having separate types that are clearly delimited. What I think we can do is have a trait implemented for both the `Response` and `Error` that can extract the unreliable peers which would make the code looking at the response type potentially more nicer. WDYT?

EdHastingsCasperAssociation commented on node/src/components/block_synchronizer/peer_list/tests.rs:
```diff
@@ -0,0 +1,118 @@
+    assert!(peer_list.is_peer_reliable(&test_peer));
+}
+
+#[test]
```
might as well add the unreliable_peer_remains_unreliable_if_demoted test case

alsrdn replied:
Added in https://github.com/casper-network/casper-node/pull/3594/commits/baa3461a4c68e9ec8e04d791e84710485fa18c9e

EdHastingsCasperAssociation reviewed (approved):
a nit and a suggestion to consider, but LGTM, gj

EdHastingsCasperAssociation commented:
bors try

alsrdn committed:
peer_list: add additional unit test for unreliable peers

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn commented:
bors try

alsrdn commented:
bors merge
